File: flink-demo/flinkScalaCode/src/main/scala/com/yw/flink/example/scalacases/case14_tableAndStreamAndSql/Case12_TimeZone.scala
package com.yw.flink.example.scalacases.case14_tableAndStreamAndSql

import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.table.api.bridge.scala.StreamTableEnvironment

import java.time.ZoneId

/**
  * Flink Table API 及SQL 编程中的时区和时间测试
  */
object Case12_TimeZone {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment

    val tableEnv: StreamTableEnvironment = StreamTableEnvironment.create(env)

    //设置时区
    tableEnv.getConfig.setLocalTimeZone(ZoneId.of("UTC"))

    val table = tableEnv.sqlQuery("select TIMESTAMP '1970-01-01 00:00:01.001' AS ntz,TO_TIMESTAMP_LTZ(4001,3) as ltz")

    tableEnv.executeSql("select CAST(ntz AS TIMESTAMP(6)) AS ntz_6, CAST(ltz AS TIMESTAMP_LTZ(6)) AS ltz_6 from " + table)
      .print()
  }
}

File: flink-demo/flinkJavaCode/src/main/java/com/yw/flink/example/javacases/case14_tableAndStreamAndSql/Case12_TimeZone.java
package com.yw.flink.example.javacases.case14_tableAndStreamAndSql;

import org.apache.flink.table.api.EnvironmentSettings;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.TableEnvironment;

import java.time.ZoneId;

/**
 * Flink Table API 及SQL 编程中的时区和时间测试
 */
public class Case12_TimeZone {
    public static void main(String[] args) {
        TableEnvironment tableEnv = TableEnvironment.create(
                EnvironmentSettings.newInstance().inStreamingMode().build());

        //设置时区
        tableEnv.getConfig().setLocalTimeZone(ZoneId.of("UTC"));
//        tableEnv.getConfig().setLocalTimeZone(ZoneId.of("Asia/Shanghai"));

        //测试时间类型
        Table table = tableEnv.sqlQuery("select " +
                "TIMESTAMP '1970-01-01 00:00:01.001' AS ntz," +
                "TO_TIMESTAMP_LTZ(4001,3) as ltz");

//        table.printSchema();
//        table.execute().print();

        tableEnv.executeSql("" +
                "select " +
                " CAST(ntz AS TIMESTAMP(6)) AS ntz_6," +
                " CAST(ltz AS TIMESTAMP_LTZ(6)) AS ltz_6" +
                " from " + table).print();
    }
}
